File: subfile1/inject_naming_glyphs.py
#!/usr/bin/env python3
# M4b: 往 file59 sub1(JIS 命名字库) 注入简体字形。
# 槽位分配钉死在下表(NAMING_EXT)——命名界面文本不进存档,但布局稳定省心。
# 渲染参数与 inject_chinese_font.render_char 完全一致(fusion-pixel 12px, 阈值64, LSB行优先)。
# 本脚本只产出 bitmap JSON,真正写盘(解压→改→compress_to_size→回写)由 inject_naming_glyphs.mjs 做。
# 用法: python3 subfile1/inject_naming_glyphs.py && node subfile1/inject_naming_glyphs.mjs
import json, os
from PIL import Image, ImageDraw, ImageFont

os.chdir(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
TTF = ImageFont.truetype('fusion-pixel-12px.otf', 12)

# 简体字 → sub1 槽位。压缩预算只有 ~103B(原tc 55437 - 最优重压 55334):
#   牺牲槽(有字形但 file61 全文从未引用, 字形换字形几乎不胀): 3224,2897,2896,2895,2894,2892,164
#   空白槽(全零, 每字 ~19B 压缩膨胀, 只能用 3 个): 3575,3574,3573
NAMING_EXT = {
    '为': 3224, '换': 2897, '汉': 2896, '确': 2895, '请': 2894, '输': 2892, '这': 164,
    '选': 3575, '假': 3574, '个': 3573,
}

def render_char(ch):
    img = Image.new('L', (12, 12), 0)
    draw = ImageDraw.Draw(img)
    bbox = draw.textbbox((0, 0), ch, font=TTF)
    w = bbox[2] - bbox[0]; h = bbox[3] - bbox[1]
    x = (12 - w) // 2 - bbox[0]
    y = (12 - h) // 2 - bbox[1]
    draw.text((x, y), ch, fill=255, font=TTF)
    bits = [1 if p > 64 else 0 for p in img.getdata()]
    out = bytearray(18)
    for i, b in enumerate(bits):
        if b: out[i // 8] |= (1 << (i % 8))
    return bytes(out)

payload = {str(slot): list(render_char(ch)) for ch, slot in NAMING_EXT.items()}
json.dump(payload, open('subfile1/naming_glyphs.json', 'w'))
json.dump({str(s): c for c, s in NAMING_EXT.items()},
          open('subfile1/codetable_jis_ext.json', 'w', encoding='utf-8'), ensure_ascii=False)
print(f'✅ 渲染 {len(payload)} 字形 → naming_glyphs.json + codetable_jis_ext.json')

File: subfile1/inject_naming_glyphs.mjs
// M4b 写盘半场: 把 naming_glyphs.json 的字形写进 WORK file59 sub1。
// sub1 在 D1 layout 的 file59 偏移 65536, 重压必须 compress_to_size 精确填满原 tc
// (与 sub0/归档同一条铁律), round-trip 校验后回写 + 修 ECC。幂等(重跑覆盖同样字节)。
// 用法: node subfile1/inject_naming_glyphs.mjs
import * as fs from 'fs';
import * as lzss from '../lib/lzss.mjs';
import { spawnSync } from 'child_process';
import { fileURLToPath } from 'url';
import * as path from 'path';
process.chdir(path.join(path.dirname(fileURLToPath(import.meta.url)), '..'));

const SECTOR = 0x930, BO = 24, BS = 2048;
const WORK = '/home/mark/Code/RomHacking/Game/P2IS_PSX/ogd/Persona 2 - Tsumi - Innocent Sin (Japan).bin';
const SUB1_OFF = 65536;   // D1 32-sector layout

const glyphs = JSON.parse(fs.readFileSync('subfile1/naming_glyphs.json', 'utf8'));
const fd = fs.openSync(WORK, 'r+');
function rblk(blk, sz) {
  const o = Buffer.allocUnsafe(sz), s = Buffer.alloc(SECTOR);
  let off = 0, sec = blk * SECTOR;
  while (off < sz) { fs.readSync(fd, s, 0, SECTOR, sec); const c = Math.min(BS, sz - off); s.copy(o, off, BO, BO + c); off += c; sec += SECTOR; }
  return o;
}
const fpd = rblk(0x17, 0x2400);
const blk = fpd.readUInt32LE(59 * 8), sz = fpd.readUInt32LE(59 * 8 + 4);
const a = rblk(blk, sz);
const tag = a.readUInt32LE(SUB1_OFF), tc = a.readUInt32LE(SUB1_OFF + 4), uc = a.readUInt32LE(SUB1_OFF + 8);
const d = Buffer.from(lzss.decompress(a, SUB1_OFF + 12, tc - 12, uc));
let n = 0;
for (const [slot, bytes] of Object.entries(glyphs)) {
  const off = 0x480 + parseInt(slot) * 18;
  if (off + 18 > d.length) { console.log(`✗ 槽 ${slot} 越界`); process.exit(1); }
  Buffer.from(bytes).copy(d, off); n++;
}
const r = lzss.compress_to_size(d, 12, tc);
if (!r) { console.log('✗ compress_to_size 装不下原 tc'); process.exit(1); }
r.writeUInt32LE(tag, 0); r.writeUInt32LE(tc, 4); r.writeUInt32LE(uc, 8);
const chk = lzss.decompress(r, 12, tc - 12, uc);
if (Buffer.compare(chk, d) !== 0) { console.log('✗ round-trip 失败'); process.exit(1); }
r.copy(a, SUB1_OFF);
let wo = 0, wsec = blk * SECTOR;
while (wo < sz) { const c = Math.min(BS, sz - wo); fs.writeSync(fd, a, wo, c, wsec + BO); wo += c; wsec += SECTOR; }
fs.closeSync(fd);
spawnSync('python3', ['fix_ecc.py', String(blk), String(blk + Math.ceil(sz / BS) - 1)], { stdio: 'ignore' });
console.log(`✅ ${n} 个简体字形注入 sub1, tc 精确填满 ${tc}, round-trip ✓`);
